### PR TITLE
feat: implement compression encoding for columar bytes values

### PR DESCRIPTION
**Pull request by ShiKaiWi**
## Rationale
Current encoding of columnar bytes is naive, without any consideration about compression.

## Detailed Changes
Introduce the compression encoding based on a designed threshold for columnar bytes.

## Test Plan
New unit tests.

### CODE REVIEW

File: components/codec/src/columnar/bytes.rs [hunk 1]
```diff
@@ -12,37 +12,119 @@
...
+        let compression =
+            Encoding::decide_compression(data_block_len, self.bytes_compress_threshold);
+        match compression {
+            Compression::NoCompression => {
+                for v in values {
+                    buf.put_slice(v);
+                }
+            }
+            Compression::Lz4 => encode_with_compress(buf, values).context(Compress)?,
         }
```
Comment (Rachelint) [inline components/codec/src/columnar/bytes.rs:112]: How about just put the these logic into `Encoding` ?

File: components/codec/src/columnar/bytes.rs [hunk 2]
```diff
@@ -52,19 +134,90 @@ impl<'a> ValuesEncoder<&'a [u8]> for ValuesEncoderImpl {
...
+    reused_buf: &mut Vec<u8>,
+    f: F,
+) -> Result<()>
+where
+    F: FnMut(Bytes) -> Result<()>,
+{
+    let mut decoder = Lz4Decoder::new(compressed_data_block_buf);
+    decoder.read_to_end(reused_buf).context(Decompress)?;
+    decode_without_compression(&mut length_block_buf, &reused_buf[..], f)
+}
```
Comment (Rachelint) [inline components/codec/src/columnar/bytes.rs:184]: Why not defining as `Encoding`'s internal method? And just expose the `encode` method?

File: components/codec/src/columnar/bytes.rs [hunk 3]
```diff
@@ -12,37 +12,119 @@
...
     {
-        for v in values {
-            debug_assert!(v.len() < u32::MAX as usize);
+        // Encode the `version`.
+        buf.put_u8(Encoding::VERSION);
+
+        // Encode the `length_block`.
+        let mut data_block_len = 0;
+        let mut length_block_len = 0;
+        for v in values.clone() {
```
Comment (jiacai2050) [inline components/codec/src/columnar/bytes.rs:95]: why clone is required here?

File: components/codec/src/columnar/bytes.rs [hunk 4]
```diff
@@ -12,37 +12,119 @@
...
+/// ```
+///
+/// Currently, the `compression` has two optional values:
+/// - 0: No compression over the data block
+/// - 1: the data block will be compressed if it is too long
+///
+/// And the lengths in the `length block` are encoded in varint.
+/// And the reason to put `length_block_len` and `compression` at the footer is
+/// for friendly decode.
+struct Encoding;
```
Comment (Rachelint) [inline components/codec/src/columnar/bytes.rs:43]: How about name it DataBlockEncoding or DataEncoding?

File: components/codec/src/columnar/bytes.rs [hunk 5]
```diff
@@ -12,37 +12,119 @@
...
+/// The layout for the string/bytes:
+/// ```plaintext
+/// +-------------+--------------+------------+-----------------------+-----------------+
+/// | version(u8) | length_block | data_block | length_block_len(u32) | compression(u8) |
+/// +-------------+--------------+------------+-----------------------+-----------------+
+/// ```
+///
+/// Currently, the `compression` has two optional values:
+/// - 0: No compression over the data block
+/// - 1: the data block will be compressed if it is too long
```
Comment (jiacai2050) [inline components/codec/src/columnar/bytes.rs:38]:
```suggestion
/// - 1: the data block will be compressed using lz4 if it is too long
```

File: components/codec/src/columnar/bytes.rs [hunk 6]
```diff
@@ -52,19 +134,90 @@ impl<'a> ValuesEncoder<&'a [u8]> for ValuesEncoderImpl {
...
+        let compression_offset = chunk.len() - Encoding::COMPRESSION_SIZE;
+        let compression = Encoding::decode_compression(chunk[compression_offset])?;
+
+        let length_block_len_offset = chunk.len() - footer_len;
+        let length_block_end = {
+            let mut len_buf = &chunk[length_block_len_offset..compression_offset];
+            len_buf.get_u32() as usize + Encoding::VERSION_SIZE
+        };
+        let mut length_block = &chunk[Encoding::VERSION_SIZE..length_block_end];
+        let data_block = &chunk[length_block_end..length_block_len_offset];
```
Comment (Rachelint) [inline components/codec/src/columnar/bytes.rs:214]: Add some comments here?

File: components/codec/src/columnar/bytes.rs (hunk 3 above)
Comment (ShiKaiWi) [reply components/codec/src/columnar/bytes.rs:95]: The iterator will be consumed if not clone.

File: components/codec/src/columnar/bytes.rs (hunk 5 above)
Comment (ShiKaiWi) [reply components/codec/src/columnar/bytes.rs:38]: The data block can be compressed by not only lz4.

File: components/codec/src/columnar/bytes.rs (hunk 4 above)
Comment (ShiKaiWi) [reply components/codec/src/columnar/bytes.rs:43]: This encoding is for the whole layout, not just data block.

**Review by Rachelint — APPROVED** [review]
LGTM

**Pull request merged**